zgw_consumers/models/certificates.py

```diff
@@ -0,0 +1,99 @@
+from datetime import date, datetime
+
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from OpenSSL import SSL, crypto
+from privates.fields import PrivateMediaFileField
+
+from ..constants import CertificateTypes
+from ..mixins import DeleteFileFieldFilesMixin
+from ..utils import pretty_print_certificate_components
+
+
+def ssl_upload_to(instance, filename):
+    return "ssl_certs_keys/{d}".format(d=date.today().strftime("%Y/%m/%d"))
+
+
+class Certificate(DeleteFileFieldFilesMixin, models.Model):
+    label = models.CharField(
+        _("label"),
+        max_length=100,
+        help_text=_("Recognisable label for the certificate"),
+    )
+    type = models.CharField(
+        _("type"),
+        max_length=20,
+        choices=CertificateTypes.choices,
+        help_text=_(
+            "Is this only a certificate or is there an associated private key?"
+        ),
+    )
+    public_certificate = PrivateMediaFileField(
+        _("public certificate"),
+        help_text=_("The content of the certificate"),
+        upload_to=ssl_upload_to,
+    )
+    private_key = PrivateMediaFileField(
+        _("private key"),
+        help_text=_("The content of the private key"),
+        blank=True,
+        upload_to=ssl_upload_to,
+    )
+
+    class Meta:
+        verbose_name = _("certificate")
+        verbose_name_plural = _("certificates")
+
+    _certificate_obj = None
+    _private_key_obj = None
+
+    @property
+    def _certificate(self):
+        if not self._certificate_obj:
+            with self.public_certificate.open(mode="rb") as certificate_f:
+                self._certificate_obj = crypto.load_certificate(
+                    crypto.FILETYPE_PEM, certificate_f.read()
+                )
+        return self._certificate_obj
+
+    @property
+    def _private_key(self):
+        if not self._private_key_obj:
+            with self.private_key.open(mode="rb") as key_f:
+                self._private_key_obj = crypto.load_privatekey(
+                    crypto.FILETYPE_PEM, key_f.read()
+                )
+        return self._private_key_obj
+
+    @property
+    def expiry_date(self) -> datetime:
+        expiry = self._certificate.get_notAfter()
+        return datetime.strptime(expiry.decode("utf-8"), "%Y%m%d%H%M%SZ")
+
+    @property
+    def issuer(self):
+        issuer_x509name = self._certificate.get_issuer()
+        return pretty_print_certificate_components(issuer_x509name)
+
+    @property
+    def subject(self):
+        subject_x509name = self._certificate.get_subject()
+        return pretty_print_certificate_components(subject_x509name)
+
+    def is_valid_key_pair(self):
+        if not self.private_key:
+            return None
+
+        context = SSL.Context(SSL.TLSv1_METHOD)
```

Comment:
do we need to provde the `TLSv1_METHOD`? TLS 1.0 and 1.1 are deprecated, so if we _have_ to provide anything, I'd opt for TLS v1.2 at minimum